trezor-crypto/crypto/base58.c

```diff
@@ -21,278 +21,254 @@
-
-	b58sz = strlen(b58);
-
-	memset(outi, 0, outisz * sizeof(*outi));
-
-	// Leading zeros, just count
-	for (i = 0; i < b58sz && b58u[i] == digits[0]; ++i)
-		++zerocount;
-
-	for ( ; i < b58sz; ++i)
-	{
-		if (b58u[i] & 0x80)
-			// High-bit set on invalid digit
-			return false;
-		if (digits_map[b58u[i]] == -1)
-			// Invalid base58 digit
-			return false;
-		c = (unsigned)digits_map[b58u[i]];
-		for (j = outisz; j--; )
-		{
-			t = ((uint64_t)outi[j]) * 58 + c;
-			c = (t & 0x3f00000000) >> 32;
-			outi[j] = t & 0xffffffff;
-		}
-		if (c)
-			// Output number too big (carry to the next int32)
-			return false;
-		if (outi[0] & zeromask)
-			// Output number too big (last int32 filled too far)
-			return false;
-	}
-
-	j = 0;
-	switch (bytesleft) {
-		case 3:
-			*(binu++) = (outi[0] &   0xff0000) >> 16;
-			//-fallthrough
-		case 2:
-			*(binu++) = (outi[0] &     0xff00) >>  8;
-			//-fallthrough
-		case 1:
-			*(binu++) = (outi[0] &       0xff);
-			++j;
-			//-fallthrough
-		default:
-			break;
-	}
-
-	for (; j < outisz; ++j)
-	{
-		*(binu++) = (outi[j] >> 0x18) & 0xff;
-		*(binu++) = (outi[j] >> 0x10) & 0xff;
-		*(binu++) = (outi[j] >>    8) & 0xff;
-		*(binu++) = (outi[j] >>    0) & 0xff;
-	}
-
-	// Count canonical base58 byte count
-	binu = bin;
-	for (i = 0; i < binsz; ++i)
-	{
-		if (binu[i]) {
-			if (zerocount > i) {
-				/* result too large */
-				return false;
-			}
-			break;
-		}
-		--*binszp;
-	}
-	*binszp += zerocount;
-
-	return true;
-}
-
-bool b58tobin(void *bin, size_t *binszp, const char *b58)
-{
-	return base58_to_bin(bin, binszp, b58, b58digits_ordered, b58digits_map);
-}
-
-int base58_check(const void *bin, size_t binsz, HasherType hasher_type, const char *base58str, const char digits[])
-{
-	unsigned char buf[32];
-	const uint8_t *binc = bin;
-	unsigned i;
-	if (binsz < 4)
-		return -4;
-	hasher_Raw(hasher_type, bin, binsz - 4, buf);
-	if (memcmp(&binc[binsz - 4], buf, 4))
-		return -1;
-
-	// Check number of zeros is correct AFTER verifying checksum (to avoid possibility of accessing base58str beyond the end)
-	for (i = 0; binc[i] == '\0' && base58str[i] == digits[0]; ++i)
-	{}  // Just finding the end of zeros, nothing to do in loop
-	if (binc[i] == '\0' || base58str[i] == digits[0])
-		return -3;
-
-	return binc[0];
-}
-
-int b58check(const void *bin, size_t binsz, HasherType hasher_type, const char *base58str)
-{
-	return base58_check(bin, binsz, hasher_type, base58str, b58digits_ordered);
+typedef uint64_t b58_maxint_t;
+typedef uint32_t b58_almostmaxint_t;
+#define b58_almostmaxint_bits (sizeof(b58_almostmaxint_t) * 8)
+const b58_almostmaxint_t b58_almostmaxint_mask =
+    ((((b58_maxint_t)1) << b58_almostmaxint_bits) - 1);
+
+// Decodes a null-terminated Base58 string `b58` to binary and writes the result
+// at the end of the buffer `bin` of size `*binszp`. On success `*binszp` is set
+// to the number of valid bytes at the end of the buffer.
+bool b58tobin(void *bin, size_t *binszp, const char *b58) {
```

Comment:
b58tobin -> b58ToBin?

Reply:
Would be nicer, but not essential, does not worth introducing another deviation from upstream.